Be/hardcoded data (#46)
Ada-Developers-Academy/core-unit-2: requests-and-responses-in-flask/flask-read-hardcoded-data.md

### !end-callout

## Hello Books API

### Before This Lesson

This lesson uses the [Hello Books API](https://github.com/AdaGold/hello-books-api). 

### This Lesson's Work

In this lesson we will build our Hello Books API to fulfill the following feature:

1. As a client, I want to send a request to get all existing books, so that I can see a list of books, with their `id`, `title`, `description`, and `author` of the book.

## Getting All Books Endpoint: Preparation

Let's consider how to implement this feature:

> As a client, I want to send a request to get all existing books, so that I can see a list of books, with their `id`, `title`, and `description` of the book.

### Data

Soon we will create and manipulate data in databases in SQL. We will then learn how to create an manipulate this data through a connection with our API built in flask. Until then, we will hardcode data in our app using data structures we have already learned about.

Let's represent our data as a list of book instances. Each book dictionary should have the keys `id`, `title`, and `description`.

We need to make a `Book` class, and then instantiate multiple instances.

> ```python
> #routes.py
> 
> class Book:
>     def __init__(id, title, description):
>         self.id = id
>         self.title = title
>         self.description = description
> 
> books = [
>     Book(1, "Fictional Book Title", "A fantasy novel set in an imaginary world."),
>     Book(2, "Fictional Book Title", "A fantasy novel set in an imaginary world."),
>     Book(3, "Fictional Book Title", "A fantasy novel set in an imaginary world.")
> ] 
> ```

Reviewer: ```suggestion
<details><summary>See our solution</summary>
```python
#routes.py

class Book:
    def __init__(id, title, description):
        self.id = id
        self.title = title
        self.description = description

books = [
    Book(1, "Fictional Book Title", "A fantasy novel set in an imaginary world."),
    Book(2, "Fictional Book Title", "A fantasy novel set in an imaginary world."),
    Book(3, "Fictional Book Title", "A fantasy novel set in an imaginary world.")
] 
```
</details>
```

Reviewer: maybe we can add a spoiler tag on this so the students can try and make the class themselves, then look at our solution

Author: Thinking on this. I think it might make sense to do the spoiler alert for the hardcoded data part, and then straight up provide the code for the route given their level of experience/exposure to the different areas.

Author: I added the spoiler alert style for the hardcoded data section, and left the controller function unhidden

| HTTP Method | Endpoint |
| ----------- | -------- |
| `GET`       | `/books` |

`GET` requests do not include a request body, so no additional planning around the request body is needed.

The response we want to send back is a list of JSON objects (dictionaries) with `id`, `title`, and `description`.

The most appropriate successful response status code is `200 OK`.

| Response Status | Response Body                                                                                              |
| --------------- | ---------------------------------------------------------------------------------------------------------- |
| `200 OK`        | `[{"id": 1, "title": "Fictional Book Title", "description": "A fantasy novel set in an imaginary world"}, {"id": 2, "title": "Fictional Book Title", "description": "A fantasy novel set in an imaginary world"}, {"id": 3, "title": "Fictional Book Title", "description": "A fantasy novel set in an imaginary world"}]` |

Now that we have an idea of what our endpoint should look like, we can turn our attention to how to implement it.

Our endpoint will need to:

1. Retrieve all of the books data.

> 1. Format the books data into the appropriate structure (list of dictionaries, where each dictionary has `id`, `title`, and `description`)

Reviewer: ```suggestion
1. Format the list of Book instances into the appropriate JSON data structure (list of dictionaries, where each dictionary has `id`, `title`, and `description`)
```

Reviewer: is this correct? Are starting with instances of classes, then converting them into JSON?

1. Send back a response

### Dependencies

To make this feature, we'll work with the following objects, types, and functions:

- `Blueprint`
- `jsonify`

#### Revisiting `Blueprint` from Flask

We will import `Blueprint` into our `routes.py` file. We will use Blueprints to create a group of related routes (endpoints). In our example, we'll create a group of related `books` routes.

### Working with `jsonify` from Flask

We need to import something named `jsonify` from `flask`.

```python
from flask import jsonify
```

`jsonify` is a Flask utility function that turns its argument into JSON. We'll use `jsonify` as a way to turn a list of book dictionaries into a `Response` object.

For additional details about `jsonify`, we can refer to:

- [Flask's definition of `jsonify`](https://flask.palletsprojects.com/en/1.1.x/api/#flask.json.jsonify)

## Getting All Books Endpoint: Code

Let's update our route function in `routes.py` to include support for retrieving our models.

> ```python
> #routes.py
> 
> from app import db
> from app.models.book import Book
> from flask import Blueprint, jsonify
> 
> class Book:
>     def __init__(id, title, description):
>         self.id = id
>         self.title = title
>         self.description = description
> 
> books = [
>     Book(1, "Fictional Book Title", "A fantasy novel set in an imaginary world."),
>     Book(2, "Fictional Book Title", "A fantasy novel set in an imaginary world."),
>     Book(3, "Fictional Book Title", "A fantasy novel set in an imaginary world.")
> ] 
> 
> books_bp = Blueprint("books", __name__, url_prefix="/books")
> 
> @books_bp.route("", methods=["GET"])
> def handle_books():
>     books_response = []
>     for book in books:
>         books_response.append({
>             "id": book.id,
>             "title": book.title,
>             "description": book.description
>         })
>     return jsonify(books_response)
> ```

Reviewer: ```suggestion
<details><summary>Show our solution</summary>
```python
#routes.py

from app import db
from app.models.book import Book
from flask import Blueprint, jsonify

class Book:
    def __init__(id, title, description):
        self.id = id
        self.title = title
        self.description = description

books = [
    Book(1, "Fictional Book Title", "A fantasy novel set in an imaginary world."),
    Book(2, "Fictional Book Title", "A fantasy novel set in an imaginary world."),
    Book(3, "Fictional Book Title", "A fantasy novel set in an imaginary world.")
] 

books_bp = Blueprint("books", __name__, url_prefix="/books")

@books_bp.route("", methods=["GET"])
def handle_books():
    books_response = []
    for book in books:
        books_response.append({
            "id": book.id,
            "title": book.title,
            "description": book.description
        })
    return jsonify(books_response)
```
</details>
```

Author: I left this as unhidden to be consistent with other lessons (for now)

| `jsonify(books_response)`                           | `books_response` contains a list of book dictionaries. To turn it into a `Response` object, we pass it into `jsonify()`. This will be our common practice when returning a list of something. When we are returning strings or dictionaries, we can use `make_response`, which we will learn about later. |
| `return ...`                                        | We must return our response. By default, a response with no specified status code returns `200 OK`                                                                                                                                                       |

### !callout-info

## `Blueprint`s and Routes are Sensitive to `/`

Be sure to specify that this endpoint is `""`. When combined with the `book_bp`'s `url_prefix`, `"/books"`, we see this endpoint is a `POST` to `/books`. URIs are sensitive to the use of `/` and Flask will throw an error if we define this route as `"/"` but try to access it as `/books`. A URI ending in `/` is different from a URI that _doesn't_ end in `/`, so we should use `""`.

<br/>

<details>

> <summary>Is there a way to declare "/" as valid a route path?</summary>

Reviewer: ```suggestion
<summary>Is there a way to declare "/" as a valid route path?</summary>
```
